core/arns/src/main/java/software/amazon/awssdk/arns/ArnResource.java

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.arns;
+
+import java.util.Objects;
+import java.util.Optional;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
+
+/**
+ * An additional model within {@link Arn} that provides the Resource Type, Resource, and
+ * Resource Qualifier of an AWS Arn when those values are present and correctly formatted
+ * within an Arn.
+ *
+ * <p>
+ * If {@link #resourceType} is not present, {@link #resource} will return the entire resource
+ * as a string the same as {@link Arn#resource()}.
+ *
+ * @see Arn
+ */
+@SdkPublicApi
+public final class ArnResource implements ToCopyableBuilder<ArnResource.Builder, ArnResource> {
+
+    private final String resourceType;
+    private final String resource;
+    private final String qualifier;
+
+    private ArnResource(DefaultBuilder builder) {
+        this.resourceType = builder.resourceType;
+        this.resource = builder.resource;
```

Comment:
If we are asserting that resource is not optional, should we validate that here? Either that or change the getter to Optional. Will it ever make sense to have ArnResource.fromString("")?

Reply:
Yeah, added validation because the resource is validated in `Arn` as well.